File: frontend/types/analysis.ts
export interface AnalysisRunRequest {
  resume_id: string;
  job_description_id: string;
  force?: boolean;
}

export interface AnalysisResult {
  id: string;
  user_id: number;
  resume_id: string;
  job_description_id: string;
  similarity_score: number;
  ats_score: number;
  missing_skills: Record<string, unknown> | null;
  details: Record<string, unknown> | null;
  notes: string | null;
  created_at: string;
}

export interface AnalysisCharts {
  ats_score: number;
  similarity_score: number;
  missing_skills: string[];
  matched_skills: string[];
  skill_gap: { skill: string; present: boolean }[];
  skill_distribution: { label: string; value: number }[];
}

export interface ScoreBreakdown {
  formatting: number;
  keyword_optimization: number;
  experience_relevance: number;
  project_relevance: number;
  weighted_skill_score: number;
  category_scores: Record<string, number>;
  confidence_score: number;
  preferredSkillImpact: number;
  bonusSkillDifferentiator: number;
}

export interface KeywordGapItem {
  keyword: string;
  found_in_resume: boolean;
  importance: "High" | "Medium" | "Low" | string;
  impact_score: number;
  category: string;
  priority: "critical" | "important" | "nice_to_have";
}

export interface ExperienceAnalysis {
  required_years: number | null;
  resume_years: number | null;
  gap: number | null;
}

export interface ProjectAnalysisItem {
  project_name: string;
  relevance_score: number;
  matched_technologies: string[];
}

export interface SoftSkills {
  matched: string[];
  missing: string[];
}

export interface AnalysisDashboardResponse {
  result: AnalysisResult;
  charts: AnalysisCharts;

  ats_score: number;
  skill_match: number;
  breakdown: ScoreBreakdown;

  resume_skills: string[];
  job_skills: string[];
  matched_skills: string[];
  missing_skills: string[];
  extra_skills: string[];

  categorized_skills: Record<string, string[]>;
  weighted_scores: Record<string, number>;

  keyword_gap: KeywordGapItem[];
  experience_analysis: ExperienceAnalysis;
  project_analysis: ProjectAnalysisItem[];
  soft_skills: SoftSkills;
  ai_recommendations: string[];
  smart_recommendations: {
    recommendations: Array<{
      type: string;
      priority: string;
      title: string;
      description: string;
      actionable: string;
      impact_score: number;
    }>;
    summary: {
      critical_issues: number;
      high_priority_issues: number;
      medium_priority_issues: number;
      low_priority_issues: number;
      total_impact_score: number;
    };
  };
}

File: backend/app/schemas/analysis.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class AnalysisRunRequest(BaseModel):
    resume_id: UUID
    job_description_id: UUID
    force: bool = False


class AnalysisResultRead(BaseModel):
    id: UUID
    user_id: int
    resume_id: UUID
    job_description_id: UUID
    similarity_score: float
    ats_score: float
    missing_skills: dict | None
    details: dict | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AnalysisCharts(BaseModel):
    ats_score: float
    similarity_score: float
    missing_skills: list[str]
    matched_skills: list[str]
    skill_gap: list[dict]
    skill_distribution: list[dict]


class ScoreBreakdown(BaseModel):
    formatting: float
    keyword_optimization: float
    experience_relevance: float
    project_relevance: float


class KeywordGapItem(BaseModel):
    keyword: str
    found_in_resume: bool
    importance: str
    impact_score: float


class ExperienceAnalysis(BaseModel):
    required_years: float | None = None
    resume_years: float | None = None
    gap: float | None = None


class ProjectAnalysisItem(BaseModel):
    project_name: str
    relevance_score: float
    matched_technologies: list[str]


class SoftSkills(BaseModel):
    matched: list[str]
    missing: list[str]


class AnalysisDashboardResponse(BaseModel):
    result: AnalysisResultRead
    charts: AnalysisCharts

    ats_score: float
    skill_match: float
    breakdown: ScoreBreakdown

    resume_skills: list[str]
    job_skills: list[str]
    matched_skills: list[str]
    missing_skills: list[str]
    extra_skills: list[str]

    keyword_gap: list[KeywordGapItem]
    experience_analysis: ExperienceAnalysis
    project_analysis: list[ProjectAnalysisItem]
    soft_skills: SoftSkills
    ai_recommendations: list[str]

